### ISSUE TITLE
Too many named `firebase-iid-executor` threads are created.

### ISSUE THREAD

**Issue by Edison0716** [firebase/firebase-android-sdk#5284]
### [REQUIRED] Step 2: Describe your environment
  * Android Studio version: 2022.3.1 Patch 1
  * Firebase version: com.google.firebase:firebase-bom:31.1.1

### [REQUIRED] Step 3: Describe the problem
![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/a9a980e8-5d53-4893-a709-331826b4e278)

### 2000+ "firebase-iid-executor" threads have been created.

**Comment by argzdev** [comment]
Hi @Edison0716, thanks for reaching out. This is interesting, could you share more details how we'll able to reproduce this?

**Comment by Edison0716** [comment]
Hi argzdev, sorry, the data report is from our company's apm, now I have not figout out how to reproduce this, I just know one thing, when the user have a long time to use our app, will be easy to create more thread by firebase, too many report data can prove that, and I also explore how to reproduce it, but need your team to double check.

**Comment by argzdev** [comment]
Thanks for the extra details, @Edison0716. I'm also not sure where `processorExecutor` is being used. I'll try to investigate this further and bring this up to our engineering sync for discussion.

**Comment by Edison0716** [comment]
@argzdev Sure, our project depends on these firebase products.

![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/fa9dcc15-5a46-4a3f-b1e2-4e0154653852)

![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/04ce5047-ed6a-4e93-87a8-17d4c4dffb39)

No more initialized code for `Crashlytics`,  and we also use it to report the custom event.

```
FirebaseCrashlytics.getInstance().setCustomKey(xxx)

FirebaseCrashlytics.getInstance().log(message: Message)

FirebaseCrashlytics.getInstance().recordException(e: Exception)
```

When call the `Application.onCreate()`, we will pull `the remote config` from firebase, like this.

## Application#onCreate Flow

### Step1
We will async the remote config, and get a value by key.
```
FirebaseRemoteConfig.getInstance().apply {
                val configSettings = FirebaseRemoteConfigSettings.Builder()
                    .setMinimumFetchIntervalInSeconds(intervalTime)
                    .build()
                setConfigSettingsAsync(configSettings)
                    .addOnCompleteListener {
                        asyncUpdate()
                    }
            }

private fun FirebaseRemoteConfig.asyncUpdate() {
        if (isUpdating.compareAndSet(false, true)) {
            if (!newConfigFetched.get()) {
                try {
                    fetchAndActivate()
                        .addOnCompleteListener { task ->
                            if (task.isSuccessful && task.result) {
                                newConfigFetched.set(true)
                                for (callback in callbacks) {
                                    // catching the exception to prevent crash happens on callback
                                    try {
                                        mFirebaseRemoteConfig?.let { config ->
                                            callback.get()?.invoke(config)
                                        }
                                    } catch (e: Exception) {
                                        appLogger.e(e)
                                    }
                                }
                                callbacks.clear()
                                isUpdating.set(false)
                            }
                        }
                } catch (e: Exception) {
                    appLogger.e(e)
                }
            }
        }
    }
```

## Step2 

report the launch time cost to firebase, use this way.

```
    private fun uploadTrace(
        traceName: String,
        duration: Long,
        attributes: Map<String, String>? = null
    ) {
        val offsetClock = OffsetClock(true, duration)
        val trace = Trace(
            traceName,
            TransportManager.getInstance(),
            offsetClock,
            AppStateMonitor.getInstance(),
            GaugeManager.getInstance()
        )
        attributes?.forEach { it ->
            trace.putAttribute(it.key, it.value)
        }
        trace.start()
        offsetClock.activateOffset = false
        trace.stop()
    }
```

## Step3
Launching a service for Firebase Cloud Messaging.
```
class MyFirebaseMessagingService : FirebaseMessagingService() {
  @WorkerThread
  public void onMessageReceived(@NonNull RemoteMessage message) {
         // to receive and parse data.
   }
}
```
next will register service to manifest file.
```
        <service
            android:name=".MyFirebaseMessagingService"
            android:exported="false">
            <intent-filter>
                <action android:name="com.google.firebase.MESSAGING_EVENT" />
            </intent-filter>
        </service>
```
That's all of firebase flow.

Go so far as to reproduce this case, I'm also checking.

**Comment by Edison0716** [comment]
@argzdev By the why, we get the thread count and details that according to the system file, respectively are `/proc/self/task` and `/proc/self/status`. 

I just worried maybe some `OOM` crash report because of that.

**Comment by argz777** [comment]
Thanks for these extra details, @Edison0716. I'll try to investigate this further and see if I can bring this up to our engineering sync to get some feedback.

**Comment by argzdev** [comment]
Thanks for these extra details, @Edison0716. I'll try to investigate this further and see if I can bring this up to our engineering sync to get some feedback.

**Comment by Edison0716** [comment]
Wating your feedback ^_^.

**Comment by argzdev** [comment]
Hi @Edison0716, I used the Android profiler tool to investigate all running threads. I was able to determine that the `firebase-iid-executor` are created when receiving FCM notifications. However, my tests shows that these threads are terminated sometime after receiving the notifications. Even when sending multiple notifications around 10-20. 

Does your company's APM account for the termination of these threads? Or does it only track creation. I would think that these threads are properly disposed of once notifications are received. If not, then it would make sense why it could cause `OOM` issues. However, I'm unable to reproduce this behavior which is making it difficult for us to pinpoint what issue you're experiencing.

If you could provide us a minimal reproducible example or steps to reproduce this behavior, it'll help us out a lot in investigating this further.

**Comment by Edison0716** [comment]
@argzdev You give me a clue, I will explore which perhaps case can create more and more threads when receive FCM from our businese side, give me some time to check. Thank you.

**Comment by argzdev** [comment]
Sure, @Edison0716. Let me know if there's anything else I can help with!

**Comment by Edison0716** [comment]
HI @argzdev 
I have a special report to prove OOM.

![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/2fb3fc11-d8fe-4722-9f35-e538991e1c94)

![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/a34c4e1c-8a1a-4b22-b89e-eab51ef24efb)

![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/7e5d474f-9409-4a20-aba9-9a914faabcb8)

**Comment by Edison0716** [comment]
It seems that we are sending too many messages in one minute. Is there a problem with the thread pool?

![img_v2_aa0ee52f-ab62-4bff-96bc-7abf75965d6g](https://github.com/firebase/firebase-android-sdk/assets/8669094/8c878cc6-3f12-4b28-ace5-d6afe7549359)

![img_v2_b259e071-40a2-4795-894e-20ab09e1a8ag](https://github.com/firebase/firebase-android-sdk/assets/8669094/9172211d-9677-4b31-be09-df484e08ccc0)

**Comment by argzdev** [comment]
Ah, I'm not sure in this case, let me consult with our product engineer to get some feedback.

**Comment by Edison0716** [comment]
When users are offline, new order messages can accumulate, and there is a certain caching period. When the user starts the app or comes online, all these accumulated messages are pushed at once, leading to excessive memory allocation and causing the app to crash (OOM - Out of Memory).

**Comment by argzdev** [comment]
Got it, I'll add these details to our report as well.

**Comment by Edison0716** [comment]
@argzdev hi, do you have any feedback?

**Comment by argzdev** [comment]
Hi @Edison0716, our engineers are still investigating this. We'll reply back here once we have updates to share. Thanks!

**Comment by argzdev** [comment]
Just some updates, @Edison0716. We're exploring a few possible ways this could be fixed or worked around. However, this will not be released anytime soon since it will require a lot of decision making and these changes will need to be reviewed and approved before it is pushed through. That said, there is no timeline when this will be available. However, this is within our radar and will be prioritized depending on severity, capacity, and demand. Thanks!

**Comment by Edison0716** [comment]
got it. thank you ! @argzdev

**Comment by gsakakihara** [comment]
This should be fixed with version 23.4.0 of firebase-messaging.

**Comment by Edison0716** [comment]
@gsakakihara Let us try to upgrade it, by the way, how to fix it? I saw this
![image](https://github.com/firebase/firebase-android-sdk/assets/8669094/7821b17c-255f-4d60-a66b-0038dbaf59ac)
this is core reason?
